lib/smack/DSAWrapper.cpp

```diff
@@ -22,185 +23,96 @@ RegisterPass<DSAWrapper>
-  nodeEqs = &getAnalysis<llvm::DSNodeEquivs>();
-  TS = &getAnalysis<dsa::TypeSafety<llvm::TDDataStructures>>();
-  memcpys = collectMemcpys(M, new MemcpyCollector(nodeEqs));
-  staticInits = collectStaticInits(M);
-  module = &M;
-  return false;
-}
-
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectMemcpys(llvm::Module &M, MemcpyCollector *mcc) {
-
-  for (llvm::Module::iterator func = M.begin(), e = M.end(); func != e;
-       ++func) {
-
-    for (llvm::Function::iterator block = func->begin(); block != func->end();
-         ++block) {
-
-      mcc->visit(*block);
+  SD = &getAnalysis<sea_dsa::DsaAnalysis>().getDsaAnalysis();
+  assert(SD->kind() == sea_dsa::GlobalAnalysisKind::CONTEXT_INSENSITIVE
+         && "Currently we only want the context-insensitive sea-dsa.");
+  for (auto& f : M.functions()) {
+    if(f.hasName() && SmackOptions::isEntryPoint(f.getName())) {
+      DG = &SD->getGraph(f);
+      break;
     }
   }
-  return mcc->getMemcpys();
+
+  collectStaticInits(M);
+  collectMemCpyds(M);
+  countGlobalRefs();
+  module = &M;
+  return false;
 }
 
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectStaticInits(llvm::Module &M) {
-  std::vector<const llvm::DSNode *> sis;
+void DSAWrapper::collectStaticInits(llvm::Module &M) {
   for (GlobalVariable &GV : M.globals()) {
     if (GV.hasInitializer()) {
       if (auto *N = getNode(&GV)) {
-        sis.push_back(N);
+        staticInits.insert(N);
       }
     }
   }
-  return sis;
-}
-
-DSGraph *DSAWrapper::getGraphForValue(const Value *V) {
-  if (const Instruction *I = dyn_cast<Instruction>(V))
-    return TD->getDSGraph(*I->getParent()->getParent());
-  else if (const Argument *A = dyn_cast<Argument>(V))
-    return TD->getDSGraph(*A->getParent());
-  else if (const BasicBlock *BB = dyn_cast<BasicBlock>(V))
-    return TD->getDSGraph(*BB->getParent());
-  else if (isa<GlobalValue>(V))
-    return TD->getGlobalsGraph();
-
-  // XXX I know this looks bad, but it works for now
-  for (auto U : V->users()) {
-    return getGraphForValue(U);
-  }
-
-  llvm_unreachable("Unexpected value.");
 }
 
-int DSAWrapper::getOffset(const MemoryLocation *l) {
-  const DSGraph::ScalarMapTy &S = getGraphForValue(l->Ptr)->getScalarMap();
-  DSGraph::ScalarMapTy::const_iterator I = S.find((const Value *)l->Ptr);
-  if (I == S.end())
-    return 0;
-  if (I->second.getNode() && I->second.getNode()->isCollapsedNode())
-    return -1;
-  unsigned offset = I->second.getOffset();
-  assert(offset <= INT_MAX && "Cannot handle large offsets");
-  return (int)offset;
-}
-
-bool DSAWrapper::isMemcpyd(const llvm::DSNode *n) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  const llvm::DSNode *nn = eqs.getLeaderValue(n);
-  for (unsigned i = 0; i < memcpys.size(); i++)
-    if (memcpys[i] == nn)
-      return true;
-  return false;
+void DSAWrapper::collectMemCpyds(llvm::Module &M) {
+  for (auto& f : M) {
+    for (auto& b : f) {
```

Comment:
Why don't you use this iterator so that you have one less loop?
http://llvm.org/docs/ProgrammersManual.html#iterating-over-the-instruction-in-a-function

Reply:
I tried to use it but encountered some compilation errors. Let me try to figure it out.

Reply:
Done.